src/evo/specialtx_validation.cpp

```diff
@@ -420,57 +420,79 @@ static bool CheckProUpRevTx(const CTransaction& tx, const CBlockIndex* pindexPre
 }
 
 // LLMQ final commitment Payload
-bool VerifyLLMQCommitment(const llmq::CFinalCommitment& qfc, const CBlockIndex* pindexQuorum)
+bool VerifyLLMQCommitment(const llmq::CFinalCommitment& qfc, const CBlockIndex* pindexPrev, CValidationState& state)
 {
-    std::vector<CBLSPublicKey> allkeys;
-    for (const auto m : deterministicMNManager->GetAllQuorumMembers((Consensus::LLMQType)qfc.llmqType, pindexQuorum)) {
-        allkeys.emplace_back(m->pdmnState->pubKeyOperator.Get());
-    }
-    return qfc.Verify(allkeys);
-}
+    AssertLockHeld(cs_main);
 
-static bool CheckLLMQCommitmentTx(const CTransaction& tx, const CBlockIndex* pindexPrev, CValidationState& state)
-{
-    llmq::LLMQCommPL pl;
-    if (!GetTxPayload(tx, pl)) {
-        return state.DoS(100, false, REJECT_INVALID, "bad-qc-payload");
+    // Check version
+    if (qfc.nVersion == 0 || qfc.nVersion > llmq::CFinalCommitment::CURRENT_VERSION) {
+        return state.DoS(100, false, REJECT_INVALID, "bad-qc-quorum-version");
     }
 
-    if (pl.nVersion == 0 || pl.nVersion > llmq::LLMQCommPL::CURRENT_VERSION) {
-        return state.DoS(100, false, REJECT_INVALID, "bad-qc-version");
-    }
-
-    Optional<Consensus::LLMQParams> params = Params().GetConsensus().GetLLMQParams(pl.commitment.llmqType);
+    // Check type
+    Optional<Consensus::LLMQParams> params = Params().GetConsensus().GetLLMQParams(qfc.llmqType);
     if (params == nullopt) {
-        return state.DoS(100, false, REJECT_INVALID, "bad-qc-type");
+        return state.DoS(100, false, REJECT_INVALID, "bad-qc-quorum-type");
     }
 
-    if (!pl.commitment.VerifySizes(*params)) {
-        return state.DoS(100, false, REJECT_INVALID, "bad-qc-invalid-sizes");
+    // Check sizes
+    if (!qfc.VerifySizes(*params)) {
+        return state.DoS(100, false, REJECT_INVALID, "bad-qc-quorum-sizes");
     }
 
     if (pindexPrev) {
-        if (pl.nHeight != (uint32_t)pindexPrev->nHeight + 1) {
-            return state.DoS(100, false, REJECT_INVALID, "bad-qc-height");
-        }
-
-        if (!mapBlockIndex.count(pl.commitment.quorumHash)) {
+        // Get quorum index
+        auto it = mapBlockIndex.find(qfc.quorumHash);
+        if (it == mapBlockIndex.end()) {
             return state.DoS(100, false, REJECT_INVALID, "bad-qc-quorum-hash");
         }
-        const CBlockIndex* pindexQuorum = mapBlockIndex.at(pl.commitment.quorumHash);
+        const CBlockIndex* pindexQuorum = it->second;
         if (pindexQuorum != pindexPrev->GetAncestor(pindexQuorum->nHeight)) {
             // not part of active chain
             return state.DoS(100, false, REJECT_INVALID, "bad-qc-quorum-hash");
```

Comment:
Let's have another error code for this DoS so we can test both invalid paths.
One is from an inexistent block while the other one is from a block that is not part of the main chain.

Reply:
Sure, but better to add it (with its test) in a small follow-up PR.